### PR TITLE
Add parseArray for parsing stream of arrays

### PR DESCRIPTION
**Pull request by adithyaov**
- [x] ~Add a `MonadThrow` constraint~
- [x] Add tests for `parseArray`
- [x] Add benchmarks for `parseArray`

We need to compare `parseArray` with `parselMx'`

### CODE REVIEW

**Comment by adithyaov** [comment]
```
benchmarked Data.Parser.ParserD/o-1-space/parseArray (full)
time                 970.7 μs   (966.7 μs .. 974.2 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 967.2 μs   (963.8 μs .. 970.7 μs)
std dev              11.83 μs   (9.752 μs .. 15.17 μs)

benchmarked Data.Parser.ParserD/o-1-space/parseArray (1050)
time                 10.56 μs   (10.53 μs .. 10.59 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 10.67 μs   (10.63 μs .. 10.79 μs)
std dev              424.3 ns   (239.4 ns .. 734.5 ns)
```

**Comment by harendra-kumar** [comment]
Performance results:

Allocations:

```
Data.Parser.ParserD(allocated)
Benchmark                                                                default(0)(Bytes) default(1) - default(0)(%)
------------------------------------------------------------------------ ----------------- --------------------------

Data.Parser.ParserD/o-1-space/parseArray (bound)                                3145692.80                   -1566.22
Data.Parser.ParserD/o-1-space/parseArray (100)                                  3136051.20                   -1798.42
```

CPU Time:

```
Data.Parser.ParserD(cpuTime)
Benchmark                                                                default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------ -------------- --------------------------

Data.Parser.ParserD/o-1-space/parseArray (100)                                   604.58                    -410.59
Data.Parser.ParserD/o-1-space/parseArray (bound)                                 612.39                    -547.03
```

**Comment by harendra-kumar** [comment]
Added fold for array streams:

```
Data.Array.Stream.Foreign(cpuTime)
Benchmark                                                                                       default(μs)
----------------------------------------------------------------------------------------------- -----------
Data.Array.Stream.Foreign/o-1-space/parse (of 100)                                                   113.14
Data.Array.Stream.Foreign/o-1-space/parse (single)                                                    91.91
Data.Array.Stream.Foreign/o-1-space/fold (of 100)                                                     58.56
Data.Array.Stream.Foreign/o-1-space/fold (single)                                                     38.65
```

File: src/Streamly/Internal/Data/Array/Stream/Foreign.hs [hunk 1]
```diff
@@ -56,15 +62,21 @@ import Streamly.Internal.Data.Array.Foreign.Type (Array(..), length)
...
+import qualified Streamly.Internal.Data.Array.Foreign.Type as Array
 import qualified Streamly.Internal.Data.Array.Foreign.Mut.Type as MA
 import qualified Streamly.Internal.Data.Array.Stream.Mut.Foreign as AS
 import qualified Streamly.Internal.Data.Fold.Type as FL
+import qualified Streamly.Internal.Data.Parser.ParserD as PRD
 import qualified Streamly.Internal.Data.Stream.IsStream as S
 import qualified Streamly.Internal.Data.Stream.StreamD as D
+-- XXX Are these dependencies OK?
+import qualified Streamly.Internal.Data.Stream.StreamD.Generate as Generate
+import qualified Streamly.Internal.Data.Stream.StreamD.Nesting as Nesting
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Stream/Foreign.hs:79]: Import `StreamD.Type` here instead?

File: src/Streamly/Internal/Data/Array/Stream/Foreign.hs [hunk 2]
```diff
@@ -385,3 +397,140 @@ _toArraysOf :: (MonadIO m, Storable a)
...
+-- Inefficient stuff but probably Ok for our use case.
+{-# INLINE insertDLL #-}
+insertDLL :: a -> DLL a -> DLL a
+insertDLL a (DLL (y, a1, x)) = (DLL (y, a1, x ++ [a]))
+
+{-# INLINE_NORMAL parseD #-}
+parseD ::
+       forall m a b. (MonadIO m, Storable a)
+    => PRD.Parser m a b
+    -> D.Stream m (Array.Array a)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Stream/Foreign.hs:447]: we should perhaps import Parser/Array types unqualified, becomes a bit easier to read.

File: src/Streamly/Internal/Data/Array/Stream/Foreign.hs [hunk 3]
```diff
@@ -385,3 +397,140 @@ _toArraysOf :: (MonadIO m, Storable a)
...
+currentDLL (DLL (_, x, _)) = x
+
+-- Inefficient stuff but probably Ok for our use case.
+{-# INLINE insertDLL #-}
+insertDLL :: a -> DLL a -> DLL a
+insertDLL a (DLL (y, a1, x)) = (DLL (y, a1, x ++ [a]))
+
+{-# INLINE_NORMAL parseD #-}
+parseD ::
+       forall m a b. (MonadIO m, Storable a)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Stream/Foreign.hs:445]: We can avoid the explicit forall?

**Comment by harendra-kumar** [comment]
@adithyaov I tried the implementation without DLL with a separate case for Done 0 case and the performance comes back to normal as with DLL.

**Comment by harendra-kumar** [comment]
The issue was because of appending arrays to create a new array, if we append the streams instead it works fine even without using the `Done 0` case.

I will push the changes we can have this API as well, it might come handy in some cases.

**Comment by harendra-kumar** [comment]
@adithyaov do you want to measure if this implementation is any better than the one already committed?

**Comment by adithyaov** [comment]
See #1178

**Pull request closed without merging**